### ISSUE TITLE
[Feature Request] Add LDAP Support

### ISSUE THREAD

**Issue by mariof1** [deluan/navidrome#141]
Will it be possible to integrate Active Directory / LDAP authentication?

**Comment by deluan** [comment]
Thanks for trying Navidrome!

I may implement LDAP support, yes. But right now I'm focused on the features from the [roadmap](https://github.com/deluan/navidrome#road-map). 

That being said, depending on the demand I may implement LDAP support before other features.

**Comment by helmut72** [comment]
LDAP support will be great. I have that many services running at home, it's even useful for private family installations.

**Comment by brianclemens** [comment]
(To help gauge demand) +1 for LDAP support!

**Comment by ericgaspar** [comment]
LDAP support for Navidrome installed with YunoHost will be a plus

**Comment by ghost-of-cerberus** [comment]
Looks like Sohalt made decent progress on #590 for adding LDAP. PR looks like it's set but @deluan, you were going to take another look. Just wanted to see if there had been any progress on this as it's not on the roadmap and it appears many of the roadmap items have been completed or are close to completion.
Again, not rushing. Just curious as I am definitely interested in this feature. Thanks!

**Comment by timothysnave** [comment]
I also don't mean to rush, just wanted to voice my support for LDAP. I have a home setup like @helmut72 with a bunch of apps, and would love to have one user to rule them all.

**Comment by andrewzah** [comment]
+1 for LDAP support as well, but if SSO solutions like Keycloak (which can sync with ldap) are supported, those are preferable to plain LDAP for me.

**Comment by PoGo606** [comment]
> +1 for LDAP support as well, but if SSO solutions like Keycloak (which can sync with ldap) are supported, those are preferable to plain LDAP for me.

I second this. LDAP and/or SSO support will be great.
Don't hesitate to reach out if any beta testing is needed !

**Comment by rafaelmathieu** [comment]
+1 for LDAP to be supported. This is the only reason I am currently still using jellyfin for music.

**Comment by fockr** [comment]
I was about to suggest Navidrome as feature app for the Freedombox. LDAP is the only feature still needed to be compatible...
Did I mention I :heart: this application? Thanks for your work, it's highly appreciated!

**Comment by zvercodebender** [comment]
+1 for LDAP support

**Comment by centuryx476** [comment]
+1 support for LDAP. This would be a killer feature...

**Comment by UserThre3** [comment]
+1 for LDAP support

**Comment by Wmartin417** [comment]
+1 for LDAP support

**Comment by rachalsolutions** [comment]
+1 for LDAP support

**Comment by Chel** [comment]
+1

**Comment by overflow-ITA** [comment]
+1

**Comment by khalimerot** [comment]
+1

**Comment by gbolton2008** [comment]
+1 for LDAP and ideally other SSO authentication methods such as OpenID Connect

**Comment by mlanies** [comment]
+1 `LDAP`

**Comment by tynorton** [comment]
+1 for LDAP

**Comment by pogmommy** [comment]
Would also love to see ldap support!

**Comment by fredmorais** [comment]
+1 for LDAP support. This would be really helpful!

**Comment by Comdriver** [comment]
Take my plus one for LDAP support from a weirdo running AD at home. Already have it for my video server, would be great to use LDAP for audio too.

**Comment by renekuehl** [comment]
+1 for LDAP/SSO. Running Samba 4 AD here, with some selfhosted apps using it for authentication. Navidrome would be welcome to do so ;)

**Comment by pimauchama** [comment]
+1 for LDAP support.

**Comment by Eschguy** [comment]
+1 for LDAP support

**Comment by FrancoisDeweulf** [comment]
+2 for LDAP support

**Comment by khalimerot** [comment]
I see a lot of people posting theses +1 for LDAP, and i think i was one of them.
But event if navidrome does not support LDAP you can have an LDAP support with the help of an authentication proxy.

I personally use Authelia for this now, and it works well (at least for my usage)

I just leave this here so people wanting LDAP can search for a solution.

**Comment by bfd69** [comment]
Hello
first let me congratulate you for the great job your doing !, i tried many servers and yours is far ahead in all terms !
+1 for ldap, i'd love to use ldap in addition to header auth, i would use header auth for regular web access and openldap for subsonic api

for me there is just openldap support and artist art folder that is missing to fully migrate from ampache to navidrome.
and just a littlte tip : 
it would be nice if you add a chown on the folders in your entrypoint if uid or gid <> 0 the service start would be more fluent.

again thanks a lot for the app anf for hearing me
best regards and happy new year !

**Comment by gmsotavio** [comment]
+1 for LDAP

**Comment by alvitali** [comment]
I'd love to see support for LDAP or alternative authentication methods like header auth or even OIDC

**Comment by renekuehl** [comment]
> I'd love to see support for LDAP or alternative authentication methods like header auth or even OIDC

You can use Header Authentication (see https://www.navidrome.org/docs/usage/security/#reverse-proxy-authentication). I used it with Authelia and it worked...

**Comment by alvitali** [comment]
> You can use Header Authentication (see https://www.navidrome.org/docs/usage/security/#reverse-proxy-authentication). I used it with Authelia and it worked...

Thanks for pointing this out, I'll definitely set this up. 

As it stands right now though, that's only a minor convenience for myself and existing users, as I'd still have to create every new user by hand. What I'd love to see is the option to automatically add a new user upon login with LDAP or OIDC. Should have written my comment better!

**Comment by helmut72** [comment]
> I'd still have to create every new user by hand.

No, it will created automatically:
https://github.com/navidrome/navidrome/issues/1723

What I miss is a logout feature and at least the first created user should be an admin user, if no header with group information is supported.

**Comment by nroach44** [comment]
This is still something I'd like to see.

**Comment by renekuehl** [comment]
> This is still something I'd like to see.

I just use authentik with ldap backend and http auth. This is sufficient enough for Homelab atm.

**Comment by zvercodebender** [comment]
+1 from me again.

**Comment by mm1ke** [comment]
+1 from me too

**Comment by anselme16** [comment]
+1 LDAP would be nice

**Comment by VijayLalwani** [comment]
+1

**Comment by zhiyan114** [comment]
I'm running AD DS and would be nice if all the members at home doesn't have to memorize multiple login credentials for internal services

**Comment by kapsh** [comment]
I'd be happy to see either #141 LDAP or #858 OIDC support implemented.

**Comment by bitabyte** [comment]
Really like the project. Fast, intuitive, responsive, and minimalist. Best open source streaming self service out there. Appreciate the subsonic API and the "radio" streams feature which really comes in handy at work. Only thing keeping me from using/ installing is lack of LDAP support. Maybe someday.

**Comment by tessamerrill** [comment]
Bump!! This would be really great for my server and I love using Navidrome...

**Comment by heretic404** [comment]
+1 for LDAP support

**Comment by joestump** [comment]
I've created a soft fork at [joestump/navidrome-ldap](https://github.com/joestump/navidrome-ldap) that updates the code from the old LDAP PR https://github.com/navidrome/navidrome/pull/590 and now have it working with my local instance of LLDAP. I've tested it successfully with a few iOS clients as well.

**Comment by z411** [comment]
Thank you @joestump - I very recently implemented OpenLDAP for my services and found out Navidrome was the only service not supporting LDAP. Do you plan to keep this fork updated in the near future?

**Comment by joestump** [comment]
@z411 yes, for now. The rebase fork is a mess so I need to address that. I'll recreate the PR if upstream is open to accepting it.

**Comment by helmut72** [comment]
I would use Apache with mod_ldap in front of Navidrome, but prefer OIDC with comfortable SSO instead. Thanks to Navidrome's header based auth, nearly every authentication system works **right now**, including LDAP.

**Comment by z411** [comment]
Does external auth work for clients (API consumers) though? I rarely use the web UI myself.

**Comment by helmut72** [comment]
API key is an API key. Use a long API key that will be pasted once into your application/Subsonic client.

External auth or this LDAP feature request is about login into the web ui. Having my single password from LDAP server pasted into a Subsonic app permanently is a way insecure. How do I know if my single password is stored securly in a music player client? A password, that will be used for maybe confidental apps too. No-go.

**Comment by z411** [comment]
So it doesn't.

**Comment by helmut72** [comment]
Of course, this is technical impossible.

If this LDAP PR works with main password, it's a security flaw and I'm glad, deluan was smart enough not to accept it.

**Comment by nroach44** [comment]
That is entirely how LDAP works. Every LDAP application works this way.

**Comment by helmut72** [comment]
Sorry, I missed something: If this LDAP PR works with main password **in a client app**, it's a security flaw. For login into web-ui it's of course ok and needed.

Modern apps and even mail providers allow creating API keys/application passwords used for the mail client. For example Nextcloud provide application passwords (or API Keys or call it how you want) for your client apps, 

"Nextcloud application passwords (or app passwords) are
unique, revocable credentials generated for specific devices or apps (e.g., phone, desktop client) instead of using the main account password. They are required when 2FA is active, enhancing security by allowing specific device access to be revoked without changing the main login"

We have 2026, not 2006.

**Comment by nroach44** [comment]
> Modern apps and even mail providers allow creating API keys/application passwords used for the mail client.

Yes, but that's done entirely at the application level. That'd mean that (and I have no idea in this situation what is or isn't being done) the fork would have to allow setting up per-app-passwords.

Now you know what's extra fun? That means that the LDAP administrator can't lock the user out of the application because the application no longer asks LDAP each time. If you change the LDAP account password, the app also probably can't check that, so the per-app keys will still work. That's a big fucking deal in the enterprise environment.

LDAP is not suitable for your use-case, if you are worried about things like that should look into something more like Oauth / SAML etc which also allows things like MFA, without the app needing to deal with all of the extra complexities.

**Comment by z411** [comment]
> Modern apps and even mail providers allow creating API keys/application passwords used for the mail client. For example Nextcloud provide application passwords (or API Keys or call it how you want) for your client apps,

It just seems to me you have an issue with LDAP in general. But this is not the place to talk about it.

Navidrome's main authentication method is user/password and the clients I use (Feishin/Tempus) only support that. If we were to push for the prioritization of more modern authentication methods (which I definitely support), that's out of scope and it's something we should take to a new issue and projects (the clients).

But right now this is about syncing the existing authentication method to a main directory. Whether that's safe or not, it fully depends on the needs and usage of each administrator.

**Comment by helmut72** [comment]
> Now you know what's extra fun? That means that the LDAP administrator can't lock the user out of the application because the application no longer asks LDAP each time. If you change the LDAP account password, the app also probably can't check that, so the per-app keys will still work. That's a big fucking deal in the enterprise environment.

I don't see any problem. Of course you need a simple housekeeping cronjob, that diffs userlist of Navidrome to LDAP group of Navidrome users and delete the difference.

**Comment by helmut72** [comment]
> It just seems to me you have an issue with LDAP in general. But this is not the place to talk about it.

No, please read 3rd post of this feature request. LDAP works with Navidrome for years: the same way as any other external authentication.

**Comment by nroach44** [comment]
> that diffs userlist of Navidrome to LDAP group of Navidrome users and delete the difference.

So if I wanted to temporarily block the user from logging in (e.g. lockout, security incident, password expiry), I now also have to check if they're in a specific group, remove them from that group, and then put them back in when it's all clear?

LDAP is for reasonably simple systems. As mentioned, oauth, SAML / SASL are modern equivalents. LDAP was not designed to be used with more detailed security requirements, that's what Kerberos is for.

As z411 says, this is to track adding LDAP support to navidrome. Some issues are inherent with that - Airsonic / subsonic / ultrasonic clients store the LDAP password directly - and other than a cursory note or warning, this is not the place to discuss them.

**Comment by helmut72** [comment]
> So if I wanted to temporarily block the user from logging in (e.g. lockout, security incident, password expiry), I now also have to check if they're in a specific group, remove them from that group, and then put them back in when it's all clear?

No. With Apache as the external auth frontdoor, you can do a check if user is locked on LDAP, if access path is `/rest` or `/share`. If user is locked, Apache returns 401 to users music client app.

> LDAP is for reasonably simple systems

Yep. My IDP system (Authelia) also use LDAP for his user database.

> this is to track adding LDAP support to navidrome

LDAP is supported since years. Thanks to Navidromes external auth system, **any** administrators prefered auth system can be used. Including LDAP. Let Navidrome developer focus on great music related features.

**Comment by kgarner7** [comment]
You _do_ know that OG Subsonic had native LDAP support? It's even part of the Subsonic spec itself (https://www.subsonic.org/pages/api.jsp#:~:text=supported%20for%20LDAP-,users.,-50, for some) and can be supported natively by subsonic clients using "legacy authentication" (sending the password as part of the request).

Obviously, reverse proxy authentication exists, but basically no client will support that.

This ticket is to track LDAP support _natively_ (which by extension, will mean support for Subsonic clients _without_ extra work). Consider taking extraneous arguments _elsewhere_.

**Comment by helmut72** [comment]
> You do know that OG Subsonic had native LDAP support?

From the docs, don't know if this matter for client apps (I don't use Subsonic anymore):

<img width="536" height="41" alt="Image" src="https://github.com/user-attachments/assets/b8649388-cd8c-4bfd-af82-d7204979108c" />

> Obviously, reverse proxy authentication exists, but basically no client will support that.

Therefore you exclude the paths from reverse proxy authentication, that be accessed from client apps, but not from checking if locked on LDAP and use the built-in app password from Navidrome.

> Consider taking extraneous arguments elsewhere.

Since this issue was flooded with +1 comments for years, I wanted to show to the interested maker, that LDAP is possible and secure with external auth. I just answered to every argument, nothing more.

Not every starter is aware what's possible with external auth and can use this information to get into the right direction.

But don't worry, be happy. I have unsubscribed to this issue now.

**Comment by realnerdc0re** [comment]
+1 LDAP support so at least one could use the same user backend thats utilized with IDP/SSO logins for other services... I just played quite a lot around with the proxy authentication and while its working, it is failing if you want to use another music player frontend...

**Comment by VijayLalwani** [comment]
+1 for native LDAP support.

**Comment by oshurkov** [comment]
Oh, LDAP support - its good. Very good...

**Comment by joestump** [comment]
I've updated my LDAP fork to start mirroring tags from upstream. Additionally, I think I have a plan to address some of the concerns here. @deluan let me know if you'd be open to accepting a PR that works like this.

Plan would be:

* Introduce "app passwords" for the Subsonic API. This is a common pattern for SSO-enabled apps and should work well with other external authentication mechanisms like OIDC. Additional upside is you get per-device tokens/passwords if you want them.
* NEVER EVER store the LDAP password – only used for web auth.
* Background task periodically scans LDAP and disables/demotes/promotes users based on LDAP updates. 

I'll be updating my fork and testing this approach on my home stack in the coming weeks.

**Comment by joestump** [comment]
Okay, I've created a version that does the following:

* Syncs username/email from LDAP when you log in via the web UI.
* Locks password and username from editing. 
* Adds an alert for the user to let them know why they can't edit username and password.
* Introduced a "Security" page that allows you to create/revoke app passwords.
* Does *not* store the LDAP password in the database.
* Will check LDAP on a schedule and remove users who are disabled/missing from LDAP.
* Will check LDAP and demote/promote admins based on LDAP memberships.

You can test this out with `ghcr.io/joestump/navidrome-ldap:develop`. Attached are a few screenshots. **NOTE:** The Subsonic API does *not* work for LDAP users until they've created an app password. There is also some somewhat funky edge cases around promoting and demoting LDAP users to/from admin. 

I have tested this locally and it works fine. ⚠️🤖**AI Disclosure:** As mentioned in the README, I am maintaining this with Claude Code. I am a reasonably competent Go programmer and I have written more than my fair share of React – I review each PR and provide relevant feedback.  

<img width="1286" height="990" alt="Image" src="https://github.com/user-attachments/assets/faef517c-9561-4b01-b2f1-713788704d31" />

<img width="1286" height="990" alt="Image" src="https://github.com/user-attachments/assets/20838c17-cbef-4ac6-8947-8c7bffe4c7c8" />
